message-transport/pkg/transport/metrics/consumer.go

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2018-Present the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package metrics
+
+import (
+	"github.com/projectriff/riff/message-transport/pkg/message"
+	"github.com/projectriff/riff/message-transport/pkg/transport"
+	"encoding/json"
+	"log"
+	"time"
+)
+
+const consumerSource = "consumer"
+
+// ConsumerAggregateMetric represents the reception of a number of message from a topic by a function in a time interval.
+type ConsumerAggregateMetric struct {
+	Topic    string
+	Function string
+	Pod      string
+	Interval time.Duration
+	Count    int32
+}
+
+// NewConsumer decorates the given delegate to send consumer metrics for the given function and pod to the given topic
+// using the given metrics producer. The given pod can be any unique identifier for the pod which will use this consumer
+// and can be pod instance-specific (that is, it need not carry across when a pod is restarted).
+func NewConsumer(delegate transport.Consumer, function string, pod string, metricsTopic string, metricsProducer transport.Producer) *consumer {
+	return &consumer{
+		function:        function,
+		pod:             pod,
+		delegate:        delegate,
+		metricsTopic:    metricsTopic,
+		metricsProducer: metricsProducer,
+	}
+}
+
+func (c *consumer) createConsumerMetricMessage(topic string) message.Message {
+	metric, err := json.Marshal(ConsumerAggregateMetric{
+		Topic:    topic,
+		Function: c.function,
+		Pod:      c.pod,
+		// TODO: aggregate metrics into suitable intervals
+		Interval: time.Duration(0),
+		Count: 1,
+	})
+	if err != nil { // should never happen
+		panic(err)
+	}
+	return message.NewMessage(metric, message.Headers{sourceHeaderKey: []string{consumerSource}})
+}
+
+type consumer struct {
+	function        string
+	pod             string
+	delegate        transport.Consumer
+	metricsTopic    string
+	metricsProducer transport.Producer
+}
+
+func (c *consumer) Receive() (message.Message, string, error) {
+	// TODO: emit "ready to receive" metric function/consumer/timestamp
+	m, t, err := c.delegate.Receive()
+	if err != nil {
+		return nil, "", err
+	}
+
+	metricsErr := c.metricsProducer.Send(c.metricsTopic, c.createConsumerMetricMessage(t))
+	if metricsErr != nil {
```

Comment:
should we send `metricsErr` back?

Reply:
I wanted to avoid a temporary problem with the metrics topic from causing a function outage, as that seemed quite brittle.

But I guess we may want to take a permanent problem with the metrics topic more seriously. One way would be to return the error and cause a function outage. We could keep a count of the number of consecutive errors and return the error if we exceed a certain threshold, but that seems like overkill. If the metrics topic has a permanent problem, the autoscaler will fail to scale appropriately: it will either scale to zero when it shouldn't or it will scale to the maximum allowed value when it shouldn't. Either of these behaviours may be sufficient to indicate that something is wrong and set users looking around for logs.

Wdyt?

Reply:
At first I thought like you that we shouldn't disrupt function invocation, but as you state it later, if the scaler itself is broken, we have a serious problem...

Reply:
@ericbottard and I agreed to return the metrics error to the caller, thereby disrupting the function, but at least not overlooking a potentially serious error. This can be revisited in future if transient errors occur sufficiently frequently.

Reply:
Fixed.